zerver/lib/export.py

```diff
@@ -1055,6 +1056,68 @@ def export_uploads_and_avatars(realm: Realm, output_dir: Path) -> None:
                              output_dir=emoji_output_dir,
                              processing_emoji=True)
 
+
+def _check_key_metadata(
+        email_gateway_bot: Optional[UserProfile],
+        key: Key, processing_avatars: bool,
+        realm: Realm, user_ids: Set[int]) -> None:
+    # This can happen if an email address has moved realms
+    if 'realm_id' in key.metadata and key.metadata['realm_id'] != str(realm.id):
+        if email_gateway_bot is None or key.metadata['user_profile_id'] != str(email_gateway_bot.id):
+            raise AssertionError("Key metadata problem: %s %s / %s" % (key.name, key.metadata, realm.id))
+        # Email gateway bot sends messages, potentially including attachments, cross-realm.
+        print("File uploaded by email gateway bot: %s / %s" % (key.name, key.metadata))
+    elif processing_avatars:
+        if 'user_profile_id' not in key.metadata:
+            raise AssertionError("Missing user_profile_id in key metadata: %s" % (key.metadata,))
+        if int(key.metadata['user_profile_id']) not in user_ids:
+            raise AssertionError("Wrong user_profile_id in key metadata: %s" % (key.metadata,))
+    elif 'realm_id' not in key.metadata:
+        raise AssertionError("Missing realm_id in key metadata: %s" % (key.metadata,))
+
+
+def _get_exported_s3_record(
+        bucket_name: str,
+        key: Key,
+        processing_avatars: bool,
+        processing_emoji: bool) -> Dict[str, Union[str, int]]:
+
+    record = dict(
+        s3_path=key.name,
+        bucket=bucket_name,
+        size=key.size,
+        last_modified=key.last_modified,
+        content_type=key.content_type,
+        md5=key.md5)
+    record.update(key.metadata)
+    if processing_emoji:
+        record['file_name'] = os.path.basename(key.name)
+    # A few early avatars don't have 'realm_id' on the object; fix their metadata
+    user_profile = get_user_profile_by_id(record['user_profile_id'])
+    if 'realm_id' not in record:
+        record['realm_id'] = user_profile.realm_id
+    record['user_profile_email'] = user_profile.email
+    # Fix the record ids
+    record['user_profile_id'] = int(record['user_profile_id'])
+    record['realm_id'] = int(record['realm_id'])
+    if processing_avatars or processing_emoji:
+        record['path'] = key.name
```

Comment:
This line should have been `            filename = os.path.join(output_dir, key.name)`